core/src/test/scala/org/broadinstitute/dsde/rawls/util/LibraryPermissionsSpec.scala

```diff
@@ -0,0 +1,129 @@
+          assertResult("org.broadinstitute.dsde.rawls.util.LibraryPermissionsSupport$$anonfun$getPermissionFunction$3")(mod.getClass.getName)
+          intercept[RawlsExceptionWithErrorReport](Await.result(mod({Future.successful(RequestComplete(StatusCodes.OK))}), Duration.Inf))
+        }
+        "should not be allowed to modify discoverability" in {
+          val mod = getPermissionFunction(discoverabilityOnly, isCurator=false, canShare=false, hasCatalogOnly=true, userLevel=WorkspaceAccessLevels.NoAccess)
+          assertResult("org.broadinstitute.dsde.rawls.util.LibraryPermissionsSupport$$anonfun$getPermissionFunction$2")(mod.getClass.getName)
+          intercept[RawlsExceptionWithErrorReport](Await.result(mod({Future.successful(RequestComplete(StatusCodes.OK))}), Duration.Inf))
+        }
+        "should not be allowed to modify publish" in {
+          val mod = getPermissionFunction(publishNames, isCurator=true, canShare=false, hasCatalogOnly=true, userLevel=WorkspaceAccessLevels.NoAccess)
+          assertResult("org.broadinstitute.dsde.rawls.util.LibraryPermissionsSupport$$anonfun$getPermissionFunction$1")(mod.getClass.getName)
+          intercept[RawlsExceptionWithErrorReport](Await.result(mod({Future.successful(RequestComplete(StatusCodes.OK))}), Duration.Inf))
+        }
+      }
+      "with read and share permissions" - {
+        "should not be allowed to modify workspace" in {
+          val mod = getPermissionFunction(modifyNames, isCurator=false, canShare=true, hasCatalogOnly=false, userLevel=WorkspaceAccessLevels.Read)
+          assertResult("org.broadinstitute.dsde.rawls.util.LibraryPermissionsSupport$$anonfun$getPermissionFunction$3")(mod.getClass.getName)
+          intercept[RawlsExceptionWithErrorReport](Await.result(mod({Future.successful(RequestComplete(StatusCodes.OK))}), Duration.Inf))
+        }
+        "should be allowed to modify discoverability" in {
+          val mod = getPermissionFunction(discoverabilityOnly, isCurator=false, canShare=true, hasCatalogOnly=false, userLevel=WorkspaceAccessLevels.Read)
+          assertResult("org.broadinstitute.dsde.rawls.util.LibraryPermissionsSupport$$anonfun$getPermissionFunction$2")(mod.getClass.getName)
+          assertResult(RequestComplete(StatusCodes.OK))(Await.result(mod({Future.successful(RequestComplete(StatusCodes.OK))}), Duration.Inf))
+        }
+        "should not be allowed to modify publish" in {
+          val mod = getPermissionFunction(publishNames, isCurator=false, canShare=true, hasCatalogOnly=false, userLevel=WorkspaceAccessLevels.Read)
+          assertResult("org.broadinstitute.dsde.rawls.util.LibraryPermissionsSupport$$anonfun$getPermissionFunction$1")(mod.getClass.getName)
+          intercept[RawlsExceptionWithErrorReport](Await.result(mod({Future.successful(RequestComplete(StatusCodes.OK))}), Duration.Inf))
+        }
+      }
+    }
```

Comment:
I see why you were wondering about reflection. Checking the `$$anonfun$getPermissionFunction$...` name here does feel fragile and difficult to read.

Best I can think of is to create a true class hierarchy, where you have a base\trait `LibraryPermChecker` and multiple subclasses, e.g. `ChangePublishedChecker`, `ChangeDiscoverabilityChecker`, `ChangeMetadataChecker`, etc. - each of which would implement a `withPermissions` method. Have `getPermissionFunction` return an instance of one of the subclasses, and have `withLibraryPermissions` use that instance. So you'd have something like:
```
          val checkerToInvoke = getPermissionFunction(names, isCurator, canShare, hasCatalogOnly, userLevel)
          checkerToInvoke.withPermissions(op)
```
then, in the tests, you can have the asserts look at the actual class name (ChangePublishedChecker vs ChangeDiscoverabilityChecker), instead of looking at the anonymous function.